Firestore/core/src/firebase/firestore/local/index_free_query_engine.cc

```diff
@@ -0,0 +1,142 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/index_free_query_engine.h"
+
+#include "Firestore/core/src/firebase/firestore/core/query.h"
+#include "Firestore/core/src/firebase/firestore/local/local_documents_view.h"
+#include "Firestore/core/src/firebase/firestore/model/document.h"
+#include "Firestore/core/src/firebase/firestore/model/maybe_document.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using core::Query;
+using model::Document;
+using model::DocumentKeySet;
+using model::DocumentMap;
+using model::DocumentSet;
+using model::MaybeDocument;
+using model::MaybeDocumentMap;
+using model::SnapshotVersion;
+
+DocumentMap IndexFreeQueryEngine::GetDocumentsMatchingQuery(
+    const Query& query,
+    const SnapshotVersion& last_limbo_free_snapshot_version,
+    const DocumentKeySet& remote_keys) const {
+  HARD_ASSERT(local_documents_view_, "SetLocalDocumentsView() not called");
+
+  // Queries that match all document don't benefit from using IndexFreeQueries.
+  // It is more efficient to scan all documents in a collection, rather than to
+  // perform individual lookups.
+  if (query.MatchesAllDocuments()) {
+    return ExecuteFullCollectionScan(query);
+  }
+
+  // Queries that have never seen a snapshot without limbo free documents should
+  // also be run as a full collection scan.
+  if (last_limbo_free_snapshot_version != SnapshotVersion::None()) {
+    return ExecuteFullCollectionScan(query);
+  }
+
+  MaybeDocumentMap documents = local_documents_view_->GetDocuments(remote_keys);
+  DocumentSet previous_results = ApplyQuery(query, documents);
+
+  if (query.limit() != Query::kNoLimit &&
+      NeedsRefill(previous_results, remote_keys,
+                  last_limbo_free_snapshot_version)) {
+    return ExecuteFullCollectionScan(query);
+  }
+
+  LOG_DEBUG("Re-using previous result from %s to execute query: %s",
+            last_limbo_free_snapshot_version.ToString(), query.ToString());
+
+  // Retrieve all results for documents that were updated since the last
+  // limbo-document free remote snapshot.
+  DocumentMap updated_results =
+      local_documents_view_->GetDocumentsMatchingQuery(
+          query, last_limbo_free_snapshot_version);
+
+  // We merge `previousResults` into `updateResults`, since `updateResults` is
+  // already a DocumentMap. If a document is contained in both lists, then
+  // its contents are the same.
+  for (Document result : previous_results) {
+    updated_results = updated_results.insert(result.key(), result);
+  }
+
+  return updated_results;
+}
+
+DocumentSet IndexFreeQueryEngine::ApplyQuery(
+    const core::Query& query, const model::MaybeDocumentMap& documents) const {
```

Comment:
Nit: you can drop qualifying the namespaces (`core::` and `model::`).

Reply:
Done